fix(router): empty resolvers cancel navigation

martinsik opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

When a route has at least two resolvers (Observables) and both of them `complete` without any `next` item the `last()` operator throws an exception.

Issue Number: #24195

## What is the new behavior?

When there's at least one empty resolver then navigation is canceled.

I think there are two ways this could work. When at least one resolver is empty or when all resolvers are empty. The original issue #24195 mentions the first option and it probably makes more sense because if I'm fine with one of the resolvers producing an empty response I can handle it in the resolver with eg. `defaultIfEmpty` operator.

## Does this PR introduce a breaking change?
```
[X] Yes
[] No
```

~This shouldn't be a breaking change because it's very unlikely that anybody was relying on the previous behavior.~

## Other information

I actually made this more simple because I used only `forkJoin` and `map` operators with single `pipe()` call.

I didn't want to update docs until I have approval from somebody competent that these changes make sense.

martinsik commented:
@kara Hey! Please, is there any chance to get this merged when I fix these merge conflicts?

BovineEnthusiast commented:
Hey, any update on this?

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -1267,6 +1268,64 @@ describe('Integration', () => {
+         expect(e).toEqual(null);
+       })));
+
+    it('should not navigate when at least resolver returns empty result',
```
```suggestion
    it('should not navigate when at least one resolver returns empty result',
```

atscott reviewed (commented):
One small nit but otherwise LGTM. Could you rebase this PR to the current master so we can run tests?

martinsik commented:
@atscott I finally rebased this PR. I had to make some refactoring because Angular Router has changed a lot in the meantime.

I was was battling with bundle size of [one integration test](https://github.com/angular/angular/blob/master/integration/_payload-limits.json#L48) because I originally solved this by adding `forkJoin()` from RxJS but since that isn't used anywhere else in this package it increased its size. I rewrote it so I'm not using `forkJoin()` at all and I also replaced `reduce()` with `takeLast()` (`reduce()` is internally implemented as a combination of `scan()` and `takeLast()`) so at the end the bundle should be even smaller after my changes. However, I wasn't able to run these tests locally so I can't tell how much.

So finally this should be good to go.

atscott commented on packages/router/src/operators/resolve_data.ts:
```diff
@@ -59,20 +57,17 @@ function resolveNode(
+        if (Object.keys(data).length === keys.length) {
+          return of (data);
+        }
+        return EMPTY;
```
Could you add a quick comment about this? Something like "ensure all resolvers returned values, otherwise ... this fixes ..."

atscott commented on packages/router/src/operators/resolve_data.ts:
```diff
@@ -27,13 +26,12 @@ export function resolveData(
                   check => runResolve(
                       check.route, targetSnapshot !, paramsInheritanceStrategy, moduleInjector)),
-              reduce((_: any, __: any) => _), map(_ => t));
+              takeLast(1), map(_ => t), );
```
Great catch here. Really weird way to have been using reduce before.

martinsik replied:
Yeah, it took me a while to figure out why is it even used there :).

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -1629,6 +1630,64 @@ describe('Integration', () => {
          expect(e).toEqual(null);
        })));
 
+    it('should not navigate when all resolvers return empty result',
```
Great tests. Thanks for adding these.

atscott reviewed (commented):
Great, thanks for adding this fix! I ran a subset of the internal presubmits yesterday and they appeared to be passing. I'll need to run them all tonight just to be safe, but I think this should be pretty much good to go.

atscott commented:
[global presubmits](https://test.corp.google.com/ui#id=OCL:302724940:BASE:303123800:1585238762147:4898923b) appear to show some failures as a result of this CL. I'll have to investigate what's going on there.

martinsik commented:
I added at least a few unit tests for `resolveData` operator using marble diagrams. While doing that I realised there's one more use-case where I have multiple `canActivate` checks (I always have at least one created somewhere in preactivation). So the way it works not is that it will cancel navigation if at least one resolver in any `canActivate` checks doesn't emit.

martinsik commented:
I was finally able to run integration tests. With `cli-hello-world-lazy-rollup` test that was previously failing because of increased bundle size the difference with this PR is `8` bytes. It went from `226631` to `226639`.

atscott commented:
> I added at least a few unit tests for `resolveData` operator using marble diagrams. While doing that I realised there's one more use-case where I have multiple `canActivate` checks (I always have at least one created somewhere in preactivation). So the way it works not is that it will cancel navigation if at least one resolver in any `canActivate` checks doesn't emit.

Oh, good. You identified the internal test failure before I got a chance to tell you about it :) 

> I'm suspicious that right now `resolveData` is running all resolvers for each `canActivate` check but I'll need to test this first.

Hmmm, that does seem suspicious. That wasn't how it was working before, was it?

> I was finally able to run integration tests. With cli-hello-world-lazy-rollup test that was previously failing because of increased bundle size the difference with this PR is 8 bytes. It went from 226631 to 226639.

Yea, the bundle size tests frequently need to be updated.

atscott commented:
Hi @martinsik, I want to clarify what the current behavior is for resolvers returning `EMPTY`:

1. If all resolvers return `EMPTY`, a `NavigationError` occurs and navigation does not complete
2. If only one of two or more resolvers return `EMPTY`, the navigation _does_ complete without error

This behavior can be confirmed by patching your two tests into master and running them. I think we can agree that the inconsistency is not desirable. Making this consistent, either by having both cases cancel navigation as is done in this PR or by allowing navigation to complete in both cases, is probably considered a breaking change because there could be applications that rely on the current behavior (knowingly or otherwise). 

I've come across several cases internally where there were `CanActivate` checks that passed and then resolvers which have `catchError` and return `EMPTY`. In these cases, the current behavior is that the route is still activated and the data is `undefined`. With this PR, the navigation would instead be cancelled. It's not easy to know what the desired behavior is here. Maybe these navigations should be cancelled, but maybe the developers prefer the navigation to continue and handle the `undefined`, meaning the resolver has to be updated to return `of(undefined)` instead of `EMPTY`. Because of this, I'm concerned that we can't make this change without unknowingly breaking applications.

An argument could also be made for fixing this in a way which allows navigation to complete in all cases rather than cancelling it, though I'm on your side of the discussion there -- if a developer really wanted `undefined` data to be resolved, he/she could specifically return an observable of `undefined`.

Lastly, this behavior is worth documenting. Can you find an appropriate place in the router documentation on resolvers to add a note?

martinsik commented:
@atscott I made a demo to make sure how it works right now: https://stackblitz.com/edit/angular-tjhbbj?file=src%2Fapp%2Fapp.module.ts

- One resolver that returns `EMPTY` => won't navigate and just ends on `ResolveStart` router event.
- Two resolvers where both return `EMPTY` => throws `Error: Uncaught (in promise): EmptyError: no elements in sequence`
- Two resolvers where one returns a value and the second one returns `EMPTY` => Navigates successfully.

With this PR all combinations from above should end with `NavigationCanceled` event.

I understand what use-cases you describe. I suppose this is a breaking change after all.

While updating documentation I found this sentence:

> If it doesn't return a valid Crisis, return an empty Observable, canceling the previous in-flight navigation to the `CrisisDetailComponent`

https://github.com/angular/angular/blob/master/aio/content/guide/router.md#fetch-data-before-navigating

So it looks like canceling navigation when resolvers return `EMPTY` is desired.

atscott commented:
> > If it doesn't return a valid Crisis, return an empty Observable, canceling the previous in-flight navigation to the `CrisisDetailComponent`

Great, I'm glad you found that so we can defend this a bug fix that aligns with already-documented expectations.

I'm adding the presubmit label to indicate that I need to fix some internal tests that would fail if this were merged now. After those are resolved, I'll need to re-run presubmits one final time.

atscott commented:
Hi @martinsik, could you rebase your PR to resolve conflicts? I'm hoping this can be merged next week.

martinsik committed:
fix(router): cancel navigation when at least one resolver completes with no "next" emission

This change aligns behavior for resolvers which return EMPTY. Currently EMPTY resolvers have inconsistent behavior:
- One resolver that returns EMPTY => won't navigate and just ends on ResolveStart router event.
- Two resolvers where both return EMPTY => throws "Error: Uncaught (in promise): EmptyError: no elements in sequence"
- Two resolvers where one returns a value and the other one returns EMPTY => Navigates successfully.
With this change any EMPTY resolver will cancel navigation.

BREAKING CHANGE: Any resolver which return EMPTY will cancel navigation.
If you want to allow the navigation to continue, you will need to update the resolvers to emit
some value, (i.e. defaultIfEmpty(...), of(...), etc).
PR Close #24195

martinsik commented:
@atscott Rebased and resolved.

atscott commented:
[global presubmit](https://test.corp.google.com/ui#id=OCL:307109456:BASE:307157733:1587179163706:c0a6413d)

atscott commented:
Hi @martinsik, I've been working through more test failures that came out of this change and realized that this needs to be marked as a breaking change. Please add a note about this at the end of your commit message. Also, please expand upon the description in the commit body so readers can quickly understand what's being changed and why. Finally, please add a note (and maybe even a test) that this will affect any resolvers used in `RouterModule.forChild` - I noticed that the current behavior (without this PR) for a `forChild` route with one resolver still continues to the route because the inherited `resolve` values are `{}`. So this is similar to the case where there are multiple resolvers.

ie, it should be something like
> fix(router): cancel navigation when at least one resolver completes with no "next" emissions
>
> This change aligns behavior for resolvers which return EMPTY. Currently...With this change, xyz
>
> BREAKING CHANGE: Any resolver which return EMPTY will cancel navigation. If you want to allow the navigation to continue, you will need to update the resolvers to emit some value, (i.e. `defaultIfEmpty(...)`, `of(...)`, etc). xyz xyz

atscott commented:
@martinsik - friendly ping to update the commit message, as mentioned in the last comment. Since this is a breaking change, we'll need to get it into a major release. If we miss v10, it will need to wait for v11.

martinsik commented:
@atscott I updated the commit message and added two more tests when using `RouterModule.forChild()`

atscott commented:
[latest global presubmit](https://test.corp.google.com/ui#id=OCL:307109456:BASE:309677946:1588564414188:61147a5a) - both failures are already failing at head.

atscott commented:
caretaker note: Please merge and sync this one separately from others if possible.
